### DISCUSSION TITLE
Expected behaviour on GSIs with empty primary key composite arrays

### DISCUSSION THREAD

**Discussion by PeterBaker0** [tywalch/electrodb#292] [General]
I am new to Electro DB but relatively familiar with DynamoDB overall. 

Here is my Electro DB schema I am playing with 

```
const UrlEntity = new Entity(
    {
        model: {
            entity: "url",
            version: "1",
            service: "app",
        },
        attributes: {
            url: {
                type: "string",
                required: true,
            },
            citation: {
                type: "string",
                required: true,
            },
            description: {
                type: "string",
                required: false,
            },
            createdAt: {
                type: "number",
                default: () => Date.now(),
                // cannot be modified after created
                readOnly: true,
            },
            updatedAt: {
                type: "number",
                // watch for changes to any attribute
                watch: "*",
                // set current timestamp when updated
                set: () => Date.now(),
                readOnly: true,
            },
        },
        indexes: {
            urls: {
                pk: {
                    field: "pk",
                    composite: ["url"],
                },
            },
            byUpdated: {
                index: "gsi1pk-gsi1sk-index",
                pk: {
                    // map to your GSI Hash/Partition key
                    field: "gsi1pk",
                    composite: [],
                },
                sk: {
                    // map to your GSI Range/Sort key
                    field: "gsi1sk",
                    composite: ["updatedAt"],
                },
            },
        },
    },
    Dynamo.Configuration
);
```

I want to enable two query use cases:

- get record by "url" id
- list all records ordered by updatedAt 

My understanding based on the quote below from [the docs](https://electrodb.dev/en/modeling/indexes/)

> For PK values, the service and version values from the model are prefixed onto the key.
> 
> For SK values, the entity value from the model is prefixed onto the key.

Suggests that an empty composite array would just result in values being automatically filled for the "gsi1pk" field with the service and version values. Previous experimentation did show this behaviour when the pk in question was the table index (rather than a GSI). 

If I create an item like so: 

```
    const result = await UrlEntity.create({
        url: inputs.url,
        citation: inputs.citation,
        description: inputs.description,
    }).go();
```

I see an item created with the following values

```
{
 "pk": "$app$url_1#url_https://test.com",
 "citation": "Test",
 "createdAt": 1693958531372,
 "updatedAt": 1693958531374,
 "url": "https://test.com",
 "__edb_e__": "url",
 "__edb_v__": "1"
}
```

There is no "gsi1pk" field, meaning that a query op against this index

```
    // Sorts by latest updated
    const result = (await UrlEntity.query
        .byUpdated({})
        .go({ order: "desc" })) as QueryResponse<typeof UrlEntity>;
```

returns an empty result.

Is this expected behaviour? 

Is there a preferred pattern for this simple set of query patterns? 

Thanks

**Comment by PeterBaker0** [comment]
I also have a minimal example to show the issue (though it may just be a misunderstanding on my part): [playground link](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2ATzABfbhiiSwAIkRwAxlQMwARsYDc481D64FAVQwI1GqtoC8YHxwLKrqWgAUomAxStGxCZKwiABccQkZsfy+mmnGqF7G+PGZCQBumLhYLnkAjEUlpTHEGGVY5nB5DNLSDU06xU0MVFZYNqhUeGmKjU0FCNOzTbF+0p0mbth8AOZ9y5nkaFjkMGlW6IP7upf75uzD1XyLVxmr68abWDt7LzGHqMc4KcwOc4DdlgMlpl4LhzNhpBoaulfjE3nlPt9wS9-oDgVwGAhiFjSpCXnC4MMgQBBKjPFEgzRrPJ8VCSGyYH6-eD4yi0sARACUYH8AD4wAARSkAOj4UBYguJ+2AwDA5gYfFlCnZYCSMCwXBwMDADC4kwwqvIlJgUOWlpgAHk+AhciCMBcbQlSVdUNIYFaaXSUWiTCy2RzFctlaxhuZOFw5KrGDs8CCoMa+NphqNxpMPaUWDG6HkQJyXlHiApzAVyOoQVI8FQGDJWHR+GAfX7Jtb6UQ4HzBcKxZLJjK5Qq8wcKQ6nS7QRHPRGvZkvvAAB5TZH7ea4QO-aQAa13PYNiGBxgPpfpzhkUCqkzSAG18oUALrzknvmJL5Y2TTuX1WkeVwrnAq55NsVS1AeAC0EFYLUuD7tBIGrpe+wHkBKJRpI3SpmAmhQAUYAAOIAMoAJJgAAEgwuB0MAAAKTC+I8YD7nAmgTssJ4IGecFQfuaG-Nesh3usD5vlxC5SbEiGYb82G4VQaYEURZGUQASuq2xwMApFyAo7GcT2CQ8XxkGIUJZKGKJ7DifkAFdjSxiST2379Iu4JesojY2Ag7yTG4xi6KIAqOKIUYMDARoMKqhiSNkhCNtshDqkaHaUmAdmSAA5EIbjDKgQhsBwxpxQI7CPKInjeOEfjxFK5KUhEyjzCkxh0CM0g7sqLB9VKgVUI1hhFKq9yIk8JiDcYOgCg12xQIK4U1T4WhSmgmDGWAUq-v+nZAi1uhzTEUoLS1cjwBgaQ5TC5g5bN9hAA)

Code here 
```
import { Entity } from "electrodb";

const UrlEntity = new Entity(
    {
        model: {
            entity: "url",
            version: "1",
            service: "app",
        },
        attributes: {
            url: {
                type: "string",
                required: true,
            },
            citation: {
                type: "string",
                required: true,
            },
            description: {
                type: "string",
                required: false,
            },
            createdAt: {
                type: "number",
                default: () => Date.now(),
                // cannot be modified after created
                readOnly: true,
            },
            updatedAt: {
                type: "number",
                // watch for changes to any attribute
                watch: "*",
                // set current timestamp when updated
                set: () => Date.now(),
                readOnly: true,
            },
        },
        indexes: {
            urls: {
                pk: {
                    field: "pk",
                    composite: ["url"],
                },
            },
            byUpdated: {
                index: "gsi1pk-gsi1sk-index",
                pk: {
                    // map to your GSI Hash/Partition key
                    field: "gsi1pk",
                    composite: [],
                },
                sk: {
                    // map to your GSI Range/Sort key
                    field: "gsi1sk",
                    composite: ["updatedAt"],
                },
            },
        },
    },
  { table: "test" }
);

// add a comment, tag, and update item's status with a condition
UrlEntity
  .create({ url:"https://www.test.com", citation: "test"})
  .go();

UrlEntity.query
  .byUpdated({ })
  .go({order: 'desc'});
```

**Comment by tywalch** [comment]
Thanks @PeterBaker0!

I have a PR out that will address this, I actually found the same issue recently, thank you for creating a discussion for this topic 👍 

I will be merging this PR soon: https://github.com/tywalch/electrodb/pull/293

**Reply by PeterBaker0** [reply]
Awesome! Thanks @tywalch

**Comment by tywalch** [comment]
@PeterBaker0 This was merged yesterday, let me know if you have any issues 👍

**Reply by PeterBaker0** [reply]
Awesome - thanks 👍

**Reply by PeterBaker0** [reply]
@tywalch I applied your update and can confirm behaviour is as expected
